### PR TITLE
Master comp jke

### PR DESCRIPTION
**Pull request by JKE-be**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_product_comparison/controllers/main.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        from_currency = request.env.user.company_id.currency_id
+        to_currency = pricelist.currency_id
+        compute_currency = lambda price: from_currency.compute(price, to_currency)
+        return compute_currency, pricelist_context
+
+    @http.route('/shop/compare/', type='http', auth="public", website=True)
+    def product_compare(self, **post):
+        values = {}
+        if post.get('products'):
+            product_ids = [int(i) for i in post.get('products').split(',') if i.isdigit()]
```
Comment (mart-e) [inline addons/website_product_comparison/controllers/main.py]: You can use ` post['products'].split(',')` here.

File: addons/website_product_comparison/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            # use search to check read access on each record/ids
+            products = request.env['product.product'].search([('id', 'in', product_ids)]).exists()
+            values['products'] = products
```
Comment (mart-e) [inline addons/website_product_comparison/controllers/main.py]: Have a nice fallback when product_ids is empty.
Going to `/shop/compare/?products=foo` is not nice and going to `/shop/compare/?products=` crash.

File: addons/website_product_comparison/static/src/js/website_product_comparison.js [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        this.update_cookie();
+        // force refresh to reposition popover
+        this.update_content(this.comparelist_product_ids, true);
+    },
+    update_cookie: function(){
+        document.cookie = 'comparelist_product_ids=' + JSON.stringify(this.comparelist_product_ids) + '; path=/';
+        this.update_comparelist_view();
+    },
+    update_comparelist_view: function() {
+        this.$('.o_product_circle').text(this.comparelist_product_ids.length);
```
Comment (mart-e) [inline addons/website_product_comparison/static/src/js/website_product_comparison.js]: This is the number of products in the cookie but not the number of displayed products
selection_010

File: addons/website_product_comparison/models/product_comparison.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+from openerp import api, fields, models
+
+
+class ProductAttributeCategory(models.Model):
+    _name = "product.attribute.category"
+    _description = "Product Attribute Category"
+    _order = 'sequence'
+
+    name = fields.Char("Category Name", required=True)
+    sequence = fields.Integer("Sequence")
```
Comment (mart-e) [inline addons/website_product_comparison/models/product_comparison.py]: `default=`

File: addons/website_product_comparison/security/ir.model.access.csv [hunk 5]
```diff
@@ -0,0 +1,6 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_product_attribute_category_public,product.attribute.category public,model_product_attribute_category,,1,0,0,0
```
Comment (mart-e) [inline addons/website_product_comparison/security/ir.model.access.csv]: Nobody has the right to create categories? Not even a manger?

File: addons/website_product_comparison/static/src/js/website_product_comparison.js [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+                });
+                self.update_cookie();
+            }
+        }
+        this.refresh_panel();
+    },
+    check_product_category: function(data) {
+        var category_ids = $('#comparelist .o_product_row:first').data('category_ids');
+        if (category_ids){
+            var categories = JSON.parse("[" + category_ids + "]"); // TO DO : use parent
```
Comment (mart-e) [inline addons/website_product_comparison/static/src/js/website_product_comparison.js]: forgotten TODO ?

File: addons/website_product_comparison/static/src/js/website_product_comparison.js [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+});
+
+var ProductComparison = Widget.extend({
+    template:"product_comparison_template",
+    events: {
+        'click .o_product_panel_header': 'toggle_panel',
+    },
+    init: function(parent){
+        this.parent = parent;
+        this.comparelist_product_ids = JSON.parse(utils.get_cookie('comparelist_product_ids') || 'null') || [];
```
Comment (mart-e) [inline addons/website_product_comparison/static/src/js/website_product_comparison.js]: `JSON.parse(utils.get_cookie('comparelist_product_ids') || '[]')` but more generally, should be fault tolerant if invalid cookie data (e.g. delete the cookie)

File: addons/website_product_comparison/static/src/xml/comparison.xml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    <t t-name="product_comparison_template">
+        <span class="o_product_panel" id="comparelist">
+            <span class="o_product_panel_header">
+                <span class="o_product_icon"><i class="fa fa-exchange" /></span>
+                <span class="o_product_text">Compare</span>
+                <span class="o_product_circle label label-primary">0</span>
+            </span>
+            <span class="o_product_panel_content">
+                <div class="o_product_panel_empty text-left">
+                    <h4>Click on <i class="fa fa-exchange" /> to add items to your comparison list</h4>
```
Comment (mart-e) [inline addons/website_product_comparison/static/src/xml/comparison.xml]: `<i />` is [invalid HTML](https://www.w3.org/TR/html5/syntax.html#start-tags), use `<i></i>` instead (yes I know we do it at many other places but it is still wrong, same for `<span />`)

File: addons/website_product_comparison/static/src/js/website_product_comparison.js [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+            container: '.o_product_feature_panel',
+            placement: 'top',
+            template: '<div style="width:600px;" class="popover comparator-popover" role="tooltip"><div class="arrow"></div><h3 class="popover-title"></h3><div class="popover-content"></div></div>',
+            content: function() {
+                return $('#comparelist .o_product_panel_content').html();
+            }
+        });
+        $('.oe_website_sale .o_add_compare, .oe_website_sale .o_add_compare_dyn').click(function (e){
+            self.parent.animate_clone($('#comparelist .o_product_panel_header'), $(this).closest('form'), -50, 10);
+            if (self.comparelist_product_ids.length < self.product_compare_limit) {
```
Comment (mart-e) [inline addons/website_product_comparison/static/src/js/website_product_comparison.js]: careful, number of products in cookies is not the same as displayed (cf screenshot below)

File: addons/website_product_comparison/static/src/js/website_product_comparison.js [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+            } else {
+                self.$('.o_comparelist_limit_warning').show();
+                self.show_panel(false);
+            }
+        });
+        self.parent.load_products(this.comparelist_product_ids).then(function() {
+            self.update_content(self.comparelist_product_ids, true);
+        });
+
+        $('body').on('click', '.comparator-popover .o_comparelist_products .o_remove', function (e){
```
Comment (mart-e) [inline addons/website_product_comparison/static/src/js/website_product_comparison.js]: ` $('.comparator-popover .o_comparelist_products .o_remove').on('click', ...` ?

**Comment by mart-e** [comment]
Also, please rename the module into `website_sale_comparison`

File: addons/website_product_comparison/controllers/main.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+            res = {}
+            for num, product in enumerate(products):
+                for var in product.attribute_line_ids:
+                    cat_name = var.attribute_id.category_id.name
```
Comment (mart-e) [inline addons/website_product_comparison/controllers/main.py]: This can be `False` (e.g. create a new attribute category) and will crash on `spec.split()` in the template

File: addons/website_product_comparison/controllers/main.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+                    att_name = var.attribute_id.name
+                    res.setdefault(cat_name, {})
```
Comment (mart-e) [inline addons/website_product_comparison/controllers/main.py]: `res.setdefault(cat_name, {}).setdefault(att_name, [' - '] * len(products))`
setdefault oneliner :fist_raised:

File: addons/website_product_comparison/controllers/main.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+                    if not res[cat_name].get(att_name):
+                        res[cat_name][att_name] = [' - '] * len(products)
+                    val = product.attribute_value_ids.filtered(lambda x: x.attribute_id == var.attribute_id)
+                    res[cat_name][att_name][num] = val and val[0].name or ' - '
```
Comment (mart-e) [inline addons/website_product_comparison/controllers/main.py]: if I understand correctly, you shall always have one and only one attribute to val, so this can be replaced by 
`res[cat_name][att_name][num] = val.name`

File: addons/website_product_comparison/models/product_comparison.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+
+class ProductAttribute(models.Model):
+    _inherit = 'product.attribute'
+
+    category_id = fields.Many2one('product.attribute.category', string="Category")
```
Comment (mart-e) [inline addons/website_product_comparison/models/product_comparison.py]: default+required would be nice but I guess you haven't found a solution for the issue of undefined record at for existing records

File: addons/website_product_comparison/models/product_comparison.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+class ProductWishlist(models.Model):
+    _name = 'product.wishlist'
+
+    partner_id = fields.Many2one('res.partner', string='Customer', required=True)
+    pricelist_id = fields.Many2one('product.pricelist', string='Pricelist when added')
+    currency_id = fields.Many2one('res.currency', related='pricelist_id.currency_id', readonly=True)
+    website_id = fields.Many2one('website', required=True)
+    price = fields.Monetary(digits=0, currency_field='currency_id')
+    price_new = fields.Float('Price', compute='compute_new_price')
```
Comment (mart-e) [inline addons/website_product_comparison/models/product_comparison.py]: These two are to differentiate the price when added to the wishlist and the current price but not something customisable? A tooltip/better string would be nice.
Is it possible to use a Monetary field too to be consistent?

File: addons/website_product_comparison/views/product_comparison_view.xml [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+        <field name="name">Website attribute categories</field>
+        <field name="type">ir.actions.act_window</field>
+        <field name="res_model">product.attribute.category</field>
+        <field name="view_mode">tree</field>
+        <field name="view_type">form</field>
+    </record>
+
+    <menuitem action="product_attribute_category_action"
+        id="menu_attribute_category_action"
+        parent="website_sale.menu_ecommerce_settings" sequence="11"/>
```
Comment (mart-e) [inline addons/website_product_comparison/views/product_comparison_view.xml]: Is this model strictly related to the ecommerce? I don't think so. `sale.prod_config_main` is maybe better (could be both though) and a different action name

File: addons/website_product_comparison/models/product_comparison.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+        res = OrderedDict()
+        for var in self.attribute_line_ids:
+            res.setdefault(var.attribute_id.category_id.name or 'Uncategorized', []).append(var)
+        return res
+
+
+class ProductWishlist(models.Model):
+    _name = 'product.wishlist'
+
+    partner_id = fields.Many2one('res.partner', string='Customer', required=True)
```
Comment (JKE-be) [inline addons/website_product_comparison/models/product_comparison.py]: owner

File: addons/website_product_comparison/static/src/js/website_product_comparison.js [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+        this.parent.show();
+    },
+    start:function(){
+        var self = this;
+        self.popover = $('#comparelist .o_product_panel_header').popover({
+            trigger: 'manual',
+            animation: true,
+            html: true,
+            title: function () {
+                return _t("Compare Products");
```
Comment (JKE-be) [inline addons/website_product_comparison/static/src/js/website_product_comparison.js]: Petite pensée pour toi: https://drive.google.com/file/d/0B1uIL9E_zXrrWUdXSHdLVDdfeUk/view

File: addons/website_product_comparison/static/src/js/website_product_comparison.js (hunk 10 above)
Comment (JKE-be) [inline addons/website_product_comparison/static/src/js/website_product_comparison.js]: anciennement, live vs click

File: addons/website_product_comparison/controllers/main.py (hunk 13 above)
Comment (JKE-be) [inline addons/website_product_comparison/controllers/main.py]: if you compare x and y, you can have 'no attribute' for the product x, but one for y

Comment (mart-e) [inline addons/website_product_comparison/controllers/main.py]: Scenario:
Prod A - attribute X, Y
Prod B - attribute Y, Z

1. Prod A,  `val=X`, res = `{'cat_X': {'X': ['val_AX', '-']}}`
2. Prod A, `val=Y`, res = `{'cat_X': {'X': ['val_AX', '-']}, 'cat_Y': {'Y': ['val_AY', '-']}}`
3. Prod B, `val=Y`, res = `{'cat_X': {'X': ['val_AX', '-']}, 'cat_Y': {'Y': ['val_AY', 'val_BY']}}`
4. Prod B, `val=Z`, res = `{'cat_X': {'X': ['val_AX', '-']}, 'cat_Y': {'Y': ['val_AY', 'val_BY']}, 'cat_Z': {'Z': ['-', 'val_BZ']},}`

`val` should always be defined

**Comment by JKE-be** [comment]
Merci ! C'est fixed ;)
